### PR TITLE
Expose waypoint and ztunnel ports

### LINKED ISSUE #45093: Waypoint and ztunnel are not exposing their ports properly
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

I'd like to create scrape_configs for ambient components, but their ports are not exposed and named properly, so I cannot do that.

This is the scrape_config I'd like to use to be aligned with the existing scrape configs recommended [here](https://istio.io/latest/docs/ops/integrations/prometheus/#option-2-customized-scraping-configurations): 

```yaml
- job_name: 'waypoint-stats'
  metrics_path: /stats/prometheus
  kubernetes_sd_configs:
  - role: pod
  relabel_configs:
  - source_labels: [__meta_kubernetes_pod_annotation_istio_io_for_service_account]
    action: keep
    regex: (.+)
  - source_labels: [__meta_kubernetes_pod_container_port_name]
    action: keep
    regex: 'waypoint-stats'
- job_name: ztunnel-stats
  kubernetes_sd_configs:
  - role: pod
  relabel_configs:
  - source_labels: [__meta_kubernetes_pod_label_app]
    action: keep
    regex: ztunnel
```

### Version

```Text
master branch
```

### Additional Information

_No response_

### Affected product area

- [X] Ambient
- [ ] Docs
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [X] Extensions and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure
- [ ] Upgrade
- [ ] Multi Cluster
- [ ] Virtual Machine
- [ ] Control Plane Revisions

### PR DESCRIPTION
**Pull request by krisztianfekete**
**Please provide a description of this PR:**

This fixes https://github.com/istio/istio/issues/45093 by exposing the ports.

### CODE REVIEW

File: manifests/charts/istio-control/istio-discovery/files/waypoint.yaml [hunk 1]
```diff
@@ -50,7 +50,17 @@ spec:
       terminationGracePeriodSeconds: 2
       serviceAccountName: {{.ServiceAccount | quote}}
       containers:
-      - args:
+      - name: istio-proxy
+        ports:
+        - containerPort: 15020
+          name: waypoint-stats
```
Comment (costinm) [inline manifests/charts/istio-control/istio-discovery/files/waypoint.yaml:56]: should it be http-... ? 

Any naming convention for prometheus ports ? What do we use in sidecars and current gateways ?

In general - I think 'waypoint' is really an internal gateway. We have a tendency to confuse users (and ourselves) with new words like waypoint or hbone - but   it's really a regular gateway using standard http connect protocol with some automated configs.

**Review by costinm — COMMENTED** [review]
Looks good - I would appreciate chaning the name of the port (not critical but we should use our own prefixes)

**Comment by costinm** [comment]
Also if you don't mind - please check if the normal gateway template has the same port, consistency is good.

File: manifests/charts/istio-control/istio-discovery/files/waypoint.yaml (hunk 1 above)
Comment (krisztianfekete) [reply manifests/charts/istio-control/istio-discovery/files/waypoint.yaml:56]: This is what we have for ingress and sidecars: https://github.com/istio/istio/blob/c014d333a18975e73e4ee358684a6b02cd51e473/manifests/charts/istio-control/istio-discovery/files/kube-gateway.yaml#L83-L88

I chose these names to attempt to make their function (and the scrape configs) easier to understand for newcomers. Let me think this trough.

Comment (howardjohn) [reply manifests/charts/istio-control/istio-discovery/files/waypoint.yaml:56]: Why don't we just put that exact thing into waypoint?

```
 - containerPort: 15021 
	   name: status-port 
	   protocol: TCP 
	 - containerPort: 15090 
	   protocol: TCP 
	   name: http-envoy-prom 
```

File: manifests/charts/istio-control/istio-discovery/files/waypoint.yaml [hunk 2]
```diff
@@ -229,10 +236,12 @@ metadata:
     uid: "{{.UID}}"
 spec:
   ports:
-  - name: https-hbone
-    port: 15008
+  {{- range $key, $val := .Ports }}
```
Comment (howardjohn) [inline manifests/charts/istio-control/istio-discovery/files/waypoint.yaml:239]: This is suspicious. Why do we need it?

Ingress exposes `status-port` on Service so Cloud LBs can health check it. Not sure why we need that for waypoint?

File: manifests/charts/istio-control/istio-discovery/files/waypoint.yaml (hunk 1 above)
Comment (krisztianfekete) [reply manifests/charts/istio-control/istio-discovery/files/waypoint.yaml:56]: Do we have anything on 15090 for waypoints?

File: manifests/charts/istio-control/istio-discovery/files/waypoint.yaml (hunk 2 above)
Comment (krisztianfekete) [reply manifests/charts/istio-control/istio-discovery/files/waypoint.yaml:239]: Good point, I can revert this so we have 15008 only.

File: manifests/charts/istio-control/istio-discovery/files/waypoint.yaml (hunk 1 above)
Comment (howardjohn) [reply manifests/charts/istio-control/istio-discovery/files/waypoint.yaml:56]: Yes ,waypoint is practically identical to ingress

Comment (kyessenov) [reply manifests/charts/istio-control/istio-discovery/files/waypoint.yaml:56]: One thing I'd like to request is to use the envoy port directly on waypoints. There's no need to go through the agent for stats (they can be O(GB) and choke the go agent).

Comment (krisztianfekete) [reply manifests/charts/istio-control/istio-discovery/files/waypoint.yaml:56]: Let me check that. Originally I got the metrics from this endpoint of the waypoint proxies: `http://localhost:15020/stats/prometheus`.

Comment (kyessenov) [reply manifests/charts/istio-control/istio-discovery/files/waypoint.yaml:56]: 15090 appears to be the envoy port that routes to admin port for stats.

Comment (howardjohn) [reply manifests/charts/istio-control/istio-discovery/files/waypoint.yaml:56]: 15090 is direct envoy, 15020 is agent+envoy+app. There is no app for waypoint, but we still do have agent metrics

File: manifests/charts/istio-control/istio-discovery/files/waypoint.yaml (hunk 2 above)
Comment (krisztianfekete) [reply manifests/charts/istio-control/istio-discovery/files/waypoint.yaml:239]: Moving @costinm's reply here:
> We probably should - don't we want Cloud LBs to connect to waypoints on
their way to the app ? I don't think it's ideal to
have Cloud LB -> Gateway -> Waypoint -> Workload.

I think exposing all would fit more use-cases, but don't have a strong opinion on it at the moment. What do you think?

**Comment by krisztianfekete** [comment]
> @costinm: Can we have 2 scrapes ? For sidecar we need the agent to proxy (to app) due
to capture and mTLS AFAIK. Even in sidecars we could get envoy directly,
I agree it's a large content and agent ends up using a lot of memory unless
it was optimized to stream.

If I am right, we always had the direct envoy recommended in our docs for both ingress and sidecars:
 - https://istio.io/latest/docs/ops/integrations/prometheus/#option-2-customized-scraping-configurations
 - https://github.com/istio/istio/blob/5b557edfa2a7f7dc44a26d25b94af0064d3ab744/manifests/charts/istio-control/istio-discovery/files/injection-template.yaml#L195C22-L197

**Comment by costinm** [comment]
On Thu, May 25, 2023 at 6:46 AM Krisztian Fekete ***@***.***>
wrote:

> @costinm <https://github.com/costinm>: Can we have 2 scrapes ? For
> sidecar we need the agent to proxy (to app) due
> to capture and mTLS AFAIK. Even in sidecars we could get envoy directly,
> I agree it's a large content and agent ends up using a lot of memory unless
> it was optimized to stream.
>
> If I am right, we always had the direct envoy recommended in our docs for
> both ingress and sidecars:
>
>    -
>    https://istio.io/latest/docs/ops/integrations/prometheus/#option-2-customized-scraping-configurations
>    -
>    https://github.com/istio/istio/blob/5b557edfa2a7f7dc44a26d25b94af0064d3ab744/manifests/charts/istio-control/istio-discovery/files/injection-template.yaml#L195C22-L197
>
>
That's option 2 and not the default, because users have app metrics as
well.

For waypoints and gateways - we can have 2 scrapes, one for agent and one
for envoy.
For sidecars - I guess we could change the code and have one scrape for
envoy and one for the agent+app metrics (i.e.
agent will not merge envoy metrics). We can move this discussion to an
issue - it shouldn't block this PR ( there
is no regression or problem added by this PR AFAIK, just a further
improvement ).

Since this PR started about Waypoints and ambient - I think it is important
to consider that for ambient we won't have
a sidecar so no agent to 'move' the app metrics to a non-intercepted port.
I'm also not sure if ztunnel is excluding 15020
from interception ( there is a bug/FR to support port exclusion ). If prom
is running in same cluster and is enabled for
ambient - it's fine, but if prom is external and doesn't have a ztunnel
sidecar - will have the same problems.

A good alternative is to recommend the use of otel-collector (or similar)
in daemon-set mode, which should work if
it is ambient-enabled.

**Comment by krisztianfekete** [comment]
> For waypoints and gateways - we can have 2 scrapes, one for agent and one
for envoy.
For sidecars - I guess we could change the code and have one scrape for
envoy and one for the agent+app metrics (i.e.
agent will not merge envoy metrics). We can move this discussion to an
issue - it shouldn't block this PR ( there
is no regression or problem added by this PR AFAIK, just a further
improvement ).

I am happy to continue this discussion in another issue. Would you mind opening one with all the questions we should address regarding this?

As per this PR, I am considering settling with just exposing the same for waypoint as what we have for gateways as well. Will push a commit shortly.

>  If prom is running in same cluster and is enabled for
ambient - it's fine, but if prom is external and doesn't have a ztunnel
sidecar - will have the same problems.

> A good alternative is to recommend the use of otel-collector (or similar)
in daemon-set mode, which should work if
it is ambient-enabled.

According to Prometheus best practices, it's always recommended running it localised, and scrape local targets, so it shouldn't matter if someone is using an OTel collector or Prometheus itself.

**Review by howardjohn — APPROVED** [review]

**Comment by costinm** [comment]
On Thu, May 25, 2023 at 8:57 AM John Howard ***@***.***>
wrote:

> We do the aggregation because users using the `prometheus.io/scrape`
> <http://prometheus.io/scrape>
> annotations can only do one (and mTLS issues). If you are writing your own
> rules it's not needed
>

It's not needed if the prometheus collector has a sidecar or client certs,
in case of sidecars.

GMP, prom operator or prom configs don't use the annotation - yet they do
have the "mTLS issues",
worse for GMP and other managed 'outside of the cluster'.

There are complicated ways to make an 'managed external' collector work
with mTLS - but I don't
know anyone doing it, instead the hack of using the annotation and injector
moving the port to something
not intercepted is effectively required.

And all this because of our (absurd IMO) insistence that sidecars are
magically 'transparent' and as such we don't
require any change in the app, everything 'just works'.

The good news is that we are now starting to accept that sidecars are not
magically transparent - and trying to
not repeat the same mistakes in ambient. Waypoints and Gateways don't have
injection or 'mTLS issues' - and
a same-node collector (otel collector for example) with ztunnel doesn't
have mTLS issues either, and I think
in ambient we should start recommending the clean solution ( since there is
no injection needed - the
prometheus.io/scrape and agent hacks won't work anyways).

My proposal for ambient would be:
- add another path for agent metrics that doesn't aggregate with envoy ( if
we don't have it already...), like /metrics/agent
- document that for sidecar and gateways user should configure prom
operator/GMP/config to scrape both agent and envoy
- leave prometheus.io/scrape pointing to envoy - that's the main app
- for apps recommend they:
    - run otel collector (or similar) per node daemon set. I would strongly
recommend this, there are many advantages
      to same-node - will write a broader do on this...
    -  add ambient to in-cluster prometheus or other collectors
    - use port 15020 for metrics ( which we can exclude from interception
or configure )
     - support client certs and mesh in external prom collectors - I doubt
that will be easy until HBONE gets broad adoption and we
    figure out how to use standards on the certs side, requiring all
collectors to natively support 'secure naming' and use Istio-specific
    spiffee certs won't work.

For sidecars - since I doubt the magical thinking will disappear before
sidecars do - we could at least document somewhere that
people who are willing to make small changes can use the separate envoy and
agent ports ( not intercepted by mTLS ) using
explicit prom configs ( they are on fixed ports, we have labels to select
pods with sidecars ). And use a fixed port with
an interception-exclude annotation for their own metrics. And people who
believe sidecars are transparent can continue
to pay the price and use the convoluted and inefficient logic we have now (
injector rewrite, merging, etc).

File: manifests/charts/istio-control/istio-discovery/files/waypoint.yaml (hunk 2 above)
Comment (hzxuzhonghu) [reply manifests/charts/istio-control/istio-discovery/files/waypoint.yaml:239]: We donot even have a service for waypoint previously, suggest supplement as needed

**Pull request merged**